lmdeploy/pytorch/model.py

```diff
@@ -0,0 +1,154 @@
+        torch.nn.init.uniform_ = self.uniform_
+        torch.nn.init.normal_ = self.normal_
+        torch.nn.init.kaiming_uniform_ = self.kaiming_uniform_
+        torch.nn.init.kaiming_normal_ = self.kaiming_normal_
+
+
+def init_model(model_path: str,
+               tokenizer_path: Optional[str] = None,
+               use_fast_tokenizer=True):
+    """Initialize model and tokenizer from given model path.
+
+    Args:
+        model_path (str): Path to model.
+        tokenizer_path (str): Path to tokenizer.
+        use_fast_tokenizer (bool): Whether to use fast tokenizer.
+
+    Note:
+        If the model is converted from new version of transformers,
+            use_fast_tokenizer should be True.
+        If using depodaca/llama-xb-hf, use_fast_tokenizer should be False.
+    """
+
+    start = time.monotonic()
+
+    if not tokenizer_path:
+        tokenizer_path = model_path
+
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path,
+                                              use_fast=use_fast_tokenizer,
+                                              trust_remote_code=True)
+
+    with LoadWoInit():
+        model = AutoModelForCausalLM.from_pretrained(model_path,
+                                                     torch_dtype=torch.float16,
+                                                     trust_remote_code=True)
+
+    logger.info(f'Model loaded in {time.monotonic() - start:.1f} seconds')
+    logger.info(f'Model loaded from {model_path}')
+    logger.debug(model)
+
+    return model, tokenizer
+
+
+def accel_model(model, accel: Optional[str] = None, max_alloc=2048, tp_size=1):
+    """Accelerate model with given accelerator.
+
+    Note:
+        Currently we support only deepspeed or just no acceleration.
+    """
+
+    logger.info(f'Accelerate model with {accel}')
+
+    if accel is None:
+        # No acceleration, just to cuda
+        # assume single gpu single process
+        # user is responsible to assign the gpu id via CUDA_VISIBLE_DEVICES # noqa: E501
+        model = model.cuda(get_local_rank())
+
+    elif accel.lower() == 'deepspeed':
+        # Use deepspeed inference inject fast kernel and/or tensor parallel
+
+        try:
+            import deepspeed
+        except ImportError as e:
+            raise ImportError('--accel=deepspeed is specified but '
+                              'deepspeed is not installed.\n'
+                              'Install with `pip install deepspeed`.') from e
+
+        config = dict(
+            tensor_parallel=dict(tp_size=tp_size),  # Use world size in general
+            dtype=torch.float16,
+            replace_with_kernel_inject=True,
+            max_out_tokens=max_alloc,
+        )
+
+        if 'InternLM' in model.__class__.__name__:
+            try:
+                # Use customized deepspeed supporting InternLM
+                # https://github.com/wangruohui/DeepSpeed/tree/support_internlm_0.10.0 (commit cdef2ce)  # noqa: E501
```

Comment:
shall we patch these codes to lmdeploy instead of in a personal repository?

Reply:
let's discussed in group.